### ISSUE TITLE
New hidden vector computation in GNNCell of PyTorch code

### ISSUE THREAD

**Issue by RuihongQiu** [CRIPAC-DIG/SR-GNN#15]
## Question 1:

The naming of variables in the [GNNCell](https://github.com/CRIPAC-DIG/SR-GNN/blob/9eff4b7fde42d891b5aa379ff4dbe0af48b2c68c/pytorch_code/model.py#L36) is a little bit frustrating such as there no inputgate in the origin paper. And the computation of the final hidden vectors is different from the origin paper.

In the paper, Eq(5) is like
![image](https://user-images.githubusercontent.com/17882988/55377455-5d299500-5558-11e9-922a-6e145c8db6c1.png)

While in the [code](https://github.com/CRIPAC-DIG/SR-GNN/blob/9eff4b7fde42d891b5aa379ff4dbe0af48b2c68c/pytorch_code/model.py#L47),

    hy = newgate + inputgate * (hidden - newgate) 

`newgate` stands for Eq(4) according to [this line](https://github.com/CRIPAC-DIG/SR-GNN/blob/9eff4b7fde42d891b5aa379ff4dbe0af48b2c68c/pytorch_code/model.py#L46) of code. `hidden` stands for the hidden vector representation of the former layers. And the `inputgate` stands for the `z` in Eq(5). It seems that code and paper are not matched.

## Question 2:

In the [computation code](https://github.com/CRIPAC-DIG/SR-GNN/blob/9eff4b7fde42d891b5aa379ff4dbe0af48b2c68c/pytorch_code/model.py#L87) of softmax along all items, why is the 0th embedding left out? After leaving out this embedding, the shape is no longer `[n_nodes, latent_size]` but `[n_nodes - 1, latent_size]`.

![image](https://user-images.githubusercontent.com/17882988/55679607-cd1d8e00-5951-11e9-8da3-8c130a0ded2b.png)

@SXKDZ @tangrizzly

**Comment by tangrizzly** [comment]
### Q1:
Thank you for pointing out. The code and the paper are not matched as for this equation. I referred to [this code](https://github.com/eecrazy/ConstructingNEEG_IJCAI_2018/blob/master/code/gnn_with_args.py#L54) when I wrote this part but did not notice this mistake and the fact that their code and paper are not matched either.
### Q2:
We use the 0th item to pad the sequences with different lengths. So when we try to generate recommendations, we need to exclude the 0th item out of the potential targets.

**Comment by RuihongQiu** [comment]
## Q1

I tested with the equation form as `Eq(5)` with `hy = hidden + inputgate * (newgate - hidden)`. The result becomes worse. It is not accord with intuition.

## Q2

So in the code,

        if opt.dataset == 'diginetica':
            n_node = 43098
        elif opt.dataset == 'yoochoose1_64' or opt.dataset == 'yoochoose1_4':
            n_node = 37484
        else:
            n_node = 310

The actual `number of item = n_node - 1`?

**Comment by tangrizzly** [comment]
Yes. `n_node` includes the padding item.
